### PR TITLE
Harshasi/fix locks

### PR DESCRIPTION
**Pull request by harsha-simhadri**
- [ ] Does this PR have a descriptive title that could go in our release notes?
- [ ] Does this PR add any new dependencies?
- [ ] Does this PR modify any existing APIs?
   - [ ] Is the change to the API backwards compatible?
- [ ] Should this result in any changes to our documentation, either updating existing docs or adding new ones?

### CODE REVIEW

File: src/index.cpp [hunk 1]
```diff
@@ -57,9 +57,10 @@ namespace diskann {
                         const bool use_opq)
       : _dist_metric(m), _dim(dim), _max_points(max_points),
         _dynamic_index(dynamic_index), _enable_tags(enable_tags),
-        _conc_consolidate(concurrent_consolidate), _query_scratch(nullptr),
-        _pq_dist(pq_dist_build), _num_pq_chunks(num_pq_chunks),
-        _use_opq(use_opq), _indexingMaxC(DEFAULT_MAXC) {
+        _indexingMaxC(DEFAULT_MAXC), _query_scratch(nullptr),
+        _conc_consolidate(concurrent_consolidate),
+        _delete_set(new tsl::robin_set<unsigned>), _pq_dist(pq_dist_build),
```
Comment (NingyuanChen) [inline src/index.cpp:62]: _delete_set(std::make_unique<tsl::robin_set<unsigned>>())

File: src/index.cpp [hunk 2]
```diff
@@ -237,16 +238,16 @@ namespace diskann {
 
   template<typename T, typename TagT>
   _u64 Index<T, TagT>::save_delete_list(const std::string &filename) {
-    if (_delete_set.size() == 0) {
+    if (_delete_set->size() == 0) {
       return 0;
     }
     std::unique_ptr<_u32[]> delete_list =
-        std::make_unique<_u32[]>(_delete_set.size());
+        std::make_unique<_u32[]>(_delete_set->size());
```
Comment (NingyuanChen) [inline src/index.cpp:245]: Lock _delete_set?

**Comment by theantony** [comment]
Shankar, Ningyuan and I reviewed together, I'm adding comments that we noted down.

File: src/index.cpp [hunk 3]
```diff
@@ -895,9 +896,6 @@ namespace diskann {
       if (pool[i].id == (unsigned) location) {
         pool.erase(pool.begin() + i);
         i--;
-      } else if (_delete_set.find(pool[i].id) != _delete_set.end()) {
```
[the remark this answers is not in the document]
Comment (harsha-simhadri) [reply src/index.cpp:898]: We want to check if the candidates returned by search have been deleted. We use _location_to_tag to keep this information.  If _location_to_tag(id) is not in the hashmap, then it has been deleted. We do this 15 lines below. So the check against delete list seems redundant and incorrect.

File: src/index.cpp [hunk 4]
```diff
@@ -1645,16 +1643,22 @@ namespace diskann {
...
+      const bool lock_adj_list, InMemQueryScratch<T> *scratch) {
+    std::vector<Neighbor> expanded_nghrs;
+    std::vector<Neighbor> pruned_list;
 
-    bool modify = false;
+    std::shared_lock<std::shared_timed_mutex> dl(_delete_lock);
 
-    for (auto ngh : _final_graph[(_u32) i]) {
+    // If this condition were not true, deadlock could result
+    assert(old_delete_set.find(loc) == old_delete_set.end());
```
[the remark this answers is not in the document]
Comment (harsha-simhadri) [reply src/index.cpp:1655]: This is really a sanity check for debugging. Can change to throw if we want to keep checking in release too.

File: src/index.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (theantony) [reply src/index.cpp:898]: I see, makes sense.

File: src/index.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (theantony) [reply src/index.cpp:1655]: Look OK to leave, there's indeed a check before calling this function.

**Comment by harsha-simhadri** [comment]
Replaced by https://github.com/microsoft/DiskANN/pull/214

**Pull request closed without merging**